parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java

```diff
@@ -1287,15 +1287,8 @@ protected <R> List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunct
          */
         final boolean containsStaleWork = wm.checkIfWorkIsStale(workContainerBatch);
```

Comment:
` wm.checkIfWorkIsStale(workContainerBatch)` is iterating over each work container and checking if it is stale - so it feels like redundant to do if we are trying to eliminate the re-scanning / re-iterating over them.

I think the filtering can happen here straight away - as we are already iterating over them anyway.
Basically the logic would look like:
 ``` 
  final Map<Boolean, List<WorkContainer<K, V>>> splitContainersMap = workContainerBatch.stream()
                .collect(Collectors.groupingBy(wm::checkIfWorkIsStale));
  final List<WorkContainer<K, V>> staleWorkContainers = splitContainersMap.getOrDefault(Boolean.TRUE, new ArrayList<>());
  if(!staleWorkContainers.isEmpty()){
      handleStaleWork(staleWorkContainers);
  }
  final List<WorkContainer<K, V>> activeWorkContainers = splitContainersMap.getOrDefault(Boolean.FALSE, new ArrayList<>());
...

```
`handleStaleWork` then accepts already filtered, only stale containers and we have `activeWorkContainers` pre-filtered and ready for processing further down.

We avoid having `handleStaleWork` method that performs hidden side effect (filtering out of stale work containers) and logic is easier to follow at a glance.

Reply:
Thanks a lot. I think this is very good idea, I miss this in the first place. Let me simplify this logic.

Reply:
fixed